Userland/Libraries/LibWeb/HTML/LazyLoadingElement.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2018-2023, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2023, Sam Atkins <atkinssj@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibWeb/DOM/Element.h>
+
+namespace Web::HTML {
+
+// Lazy-loaded elements should invoke this macro to inject overridden LazyLoadingElement methods.
+#define LAZY_LOADING_ELEMENT(ElementClass)                                     \
+private:                                                                       \
+    virtual HTMLElement& lazy_loading_element_to_html_element() override       \
+    {                                                                          \
+        static_assert(IsBaseOf<HTMLElement, ElementClass>);                    \
+        return *this;                                                          \
+    }                                                                          \
+                                                                               \
+    virtual LazyLoadingElement& dom_element_to_lazy_loading_element() override \
+    {                                                                          \
+        return *this;                                                          \
+    }
+
+enum class LazyLoading {
+    Lazy,
+    Eager,
+};
+
+class LazyLoadingElement {
+public:
+    // https://html.spec.whatwg.org/multipage/urls-and-fetching.html#lazy-loading-attributes
+    [[nodiscard]] LazyLoading lazy_loading_attribute() const;
+
+    // https://html.spec.whatwg.org/multipage/urls-and-fetching.html#will-lazy-load-element-steps
+    [[nodiscard]] bool will_lazy_load_element() const;
+
+    void set_lazy_load_resumption_steps(Function<void()>);
+    JS::GCPtr<JS::HeapFunction<void()>> take_lazy_load_resumption_steps(Badge<DOM::Document>);
+
+    void visit_lazy_loading_element(JS::Cell::Visitor&);
+
+    virtual HTMLElement& lazy_loading_element_to_html_element() = 0;
+    HTMLElement const& lazy_loading_element_to_html_element() const
+    {
+        return const_cast<LazyLoadingElement&>(*this).lazy_loading_element_to_html_element();
+    }
```

Comment:
The way this is used makes it feel like this actually wants to be a CRTP thing, rather than a class with actual virtual methods.

Reply:
Sure!